File: code/src/wang/relish/algorithm/handler/MessageQueue.java
package wang.relish.algorithm.handler;

//优先级队列
public class MessageQueue {


    //队列大小
    private int size;

    //队列头
    private Node head;

    //队列尾
    private Node tail;


    public MessageQueue() {
        this.size = 0;
        this.head = null;
        this.tail = null;
    }

    /**
     * 维持的有序列表节点
     **/
    private static class Node {
        Node pre;
        Node next;
        Message message;

        public Node(Message message) {
            this.message = message;
        }
    }


    private boolean insertToLinkTail(Node node) {

        if (this.head == null) {
            this.head = this.tail = node;
            return true;
        }

        //顺序插入
        Node p = head;
        while (p != null) {
            if (p.message.when > node.message.when) {
                break;
            }
            p = p.next;
        }
        if (p == head) {
            //头部
            p.pre = node;
            node.next = p;
            head = node;
            return true;
        } else if (p == null) {
            //尾部
            tail.next = node;
            node.pre = tail;
            tail = node;
            return true;
        } else {
            //中间
            Node pre = p.pre;
            pre.next = node;
            node.next = p;
            node.pre = pre;
            p.pre = node;
            return true;
        }
    }

    private Node removeFromLinkHead() {

        if (head == null) return null;

        Node p = head;
        Node newHead = head.next;
        if (newHead == null) {
            head = tail = null;
        } else {
            newHead.pre = null;
            head = newHead;
        }
        return p;
    }


    /**
     * 消息api
     **/

    public boolean offer(Message message) {
        synchronized (this) {
            if (this.insertToLinkTail(new Node(message))) {
                this.size++;
                return true;
            }
            return false;
        }
    }

    public Message poll() {
        synchronized (this) {
            Node node = this.removeFromLinkHead();

            if (node == null) {
                return null;
            }
            this.size--;
            return node.message;
        }
    }

    public Message peek() {
        synchronized (this) {
            if (head == null) return null;
            return head.message;
        }
    }

    public int getSize() {
        synchronized (this) {
            return this.size;
        }
    }
}

File: code/src/wang/relish/algorithm/handler/mini/MessageQueue.kt
package wang.relish.algorithm.handler.mini

import java.util.concurrent.LinkedBlockingDeque
import kotlin.math.min

class MessageQueue {


    // val queue = LinkedBlockingDeque<Message>()
    // 单链表头
    var mMessages: Message? = null

    fun enqueueMessage(msg: Message) {
        enqueueMessage(msg, 0L)
    }

    fun enqueueMessage(msg: Message, `when`: Long) {
        if (msg.target == null) {
            throw IllegalArgumentException("Message must have a target.");
        }
        msg.`when` = `when`
        var p: Message? = mMessages
        if (p == null || `when` == 0L || `when` < p.`when`) {
            msg.next = p
            mMessages = msg
        } else {
            var prev: Message?
            while (true) {
                prev = p
                p = p?.next
                if (p == null || `when` < p.`when`) {
                    break
                }
            }
            msg.next = p
            prev?.next = msg
        }
    }

    private fun last(): Message? {
        val h = mMessages
        if (h == null) return null
        var cur = h
        while (cur?.next != null) {
            cur = cur?.next
        }
        return cur
    }

    internal fun next(): Message? {
        var nextPollTimeoutMillis = 0L
        while (true) {
            nativePollOnce(nextPollTimeoutMillis)
            synchronized(this) {
                val now = System.currentTimeMillis()
                var prevMsg: Message? = null
                var msg = mMessages
                if (msg != null && now < msg.`when`) {
                    nextPollTimeoutMillis = min(msg.`when` - now, Integer.MAX_VALUE.toLong())
                } else {
                    if (prevMsg != null) {
                        prevMsg.next = msg?.next
                    } else {
                        mMessages = msg?.next
                    }
                    msg?.next = null
                    return msg
                }
            }
        }
        return null
    }

    fun nativePollOnce(nextPollTimeout: Long) {
        Thread.sleep(nextPollTimeout)
    }
}